Mathlib/Tactic/SolveByElim.lean

```diff
@@ -15,40 +18,292 @@ We'll gradually bring it up to feature parity.
 
 open Lean Meta Elab Tactic
 
-/-- Return local hypotheses which are not "implementation detail", as `Expr`s. -/
-def Lean.Meta.getLocalHyps : MetaM (Array Expr) := do
-  let mut hs := #[]
-  for d in ← getLCtx do
-    if !d.isImplementationDetail then hs := hs.push d.toExpr
-  return hs
+/-- Visualize an `Except` using a checkmark or a cross. -/
+def exceptEmoji : Except ε α → String
+  | .error _ => crossEmoji
+  | .ok _ => checkEmoji
+
+namespace Lean.MVarId
+
+/--
+`applyFirst lemmas cont goal` will try to apply one of the `lemmas` to the goal `goal`,
+and then call `cont` on the resulting `List MVarId` of subgoals.
+
+It returns the result from `cont` for the first such lemma for which
+both the `apply` and the call to `cont` succeed.
+
+``applyFirst (trace := `name)`` will construct trace nodes for ``name` indicating which
+calls to `apply` succeeded or failed.
+-/
+-- Because the operation of this function via a continuation is fairly specific to `solve_by_elim`,
+-- we keep it here rather than moving it into `Mathlib/Lean/`.
+def applyFirst (cfg : ApplyConfig := {}) (trace : Name := .anonymous) (lemmas : List Expr)
+    (cont : List MVarId → MetaM α) (g : MVarId) : MetaM α :=
+  lemmas.firstM fun e =>
+    withTraceNode trace (return m!"{exceptEmoji ·} tried to apply: {e}") do
```

Comment:
I know I argued for the past tense ("tried to apply") in #681, but I think I've come around to the present tense ("trying to apply"). My mental model now is that this is a log line with a hole that gets filled in later.